lib/praxis/resource_definition.rb

```diff
@@ -48,6 +48,37 @@ def version(version=nil, options= { using: [:header,:params] }.freeze )
         @version_options = options
       end
 
+      def canonical_path( action_name=nil )
+        if action_name
+          raise "Canonical path for #{self.name} is already defined as: '#{@canonical_action_name}'. 'canonical_path' can only be defined once." if @canonical_action_name
+          @canonical_action_name = action_name
+        else
+          # Resolution of the actual action definition needs to be done lazily, since we can use the `canonical_path` stanza
+          # at the top of the resource, well before the actual action is defined.
+          unless @canonical_action
+            href_action = @canonical_action_name || DEFAULT_RESOURCE_HREF_ACTION
+            @canonical_action = actions.fetch(href_action) do
+              raise "Error: trying to set canonical_href of #{self.name}. Action '#{href_action}' does not exist"
+            end
+          end
+          return @canonical_action
+        end
+      end
+      
+      def to_href( params )
+        canonical_path.primary_route.path.expand(params)
+      end
+
+      def parse_href(path)
```

Comment:
You're gonna hate me, but I just realized this method is really just doing this:

```ruby
def parse_href(path)
  param_values = canonical_path.primary_route.path.params(path)
  canonical_path.params.load(param_values).to_h
rescue => e
  raise Praxis::Exception.new("Error parsing or coercing parameters from href: #{path}\n"+e.message)
end
```